compiler/src/dotty/tools/dotc/transform/Inlining.scala

```diff
@@ -63,6 +63,12 @@ class Inlining extends MacroTransform with IdentityDenotTransformer {
   }
 
   private class InliningTreeMap extends TreeMapWithImplicits {
+
+    /** List of top level classes added by macro annotation in a package object.
+     *  These are added the PackageDef that owns this particular package object.
+     */
+    private val topClasses = new collection.mutable.ListBuffer[Tree]
```

Comment:
Unfortunately I think this isn't sufficient because package objects can be nested:
```scala
package foo {
  val x = 1
  package bar {
    val y = 2
  }
}
```
Instead, maybe the MemberDef case of transform should return a Thicket with the top-level classes, and we should add an extra case to `transform` to handle the package object module class itself, where we should also return a Thicket with the top-level classes

Reply:
This use case was considered and works. I added tests for it in `tests/run-macros/annot-add-global-class`.

Reply:
Note that after after post typer the tree is
```
package foo {
  package bar {
    val y = 2
  }
  val x = 1
}
```

This implies that nested classes are processed first and the buffer never overlaps and is emptied just after transforming the nested package.

Reply:
> This implies that nested classes are processed first and the buffer never overlaps and is emptied just after transforming the nested package.

This is subtle, so this precondition should be documented in the code (and ideally checked somewhere, in case it breaks)

Reply:
I found cases where this precondition does not hold. I updated the implementation to handle such cases.